File: rtl/NSString.java
package rtl;

import java.util.Locale;

public class NSString extends NSObject {

    // ------------------- Поле -------------------
    private String string;

    // ------------------- Статический ClassInfo -------------------
    public static ClassInfo s_classInfo;

    // ------------------- Конструкторы -------------------
    public NSString() {
        this.string = "";
    }

    public NSString(String str) {
        this.string = str;
    }

    public NSString(NSString other) {
        this.string = other.string;
    }

    // ------------------- Статические методы -------------------
    public static NSString stringStatic() {
        return new NSString("");
    }

    public static NSString stringWithCStringStatic(String cstr) {
        return new NSString(cstr);
    }

    public static NSString stringWithStringStatic(NSString str) {
        return new NSString(str);
    }

    // ------------------- Динамические методы -------------------
    public char characterAtIndexDynamic(int index) {
        if (index < 0 || index >= string.length()) return '\0';
        return string.charAt(index);
    }

    public int hasPrefixDynamic(NSString prefix) {
        return string.startsWith(prefix.string) ? 1 : 0;
    }

    public int hasSuffixDynamic(NSString suffix) {
        return string.endsWith(suffix.string) ? 1 : 0;
    }

    public int lengthDynamic() {
        return string.length();
    }

    public int intValueDynamic() {
        try {
            return Integer.parseInt(string);
        } catch (NumberFormatException e) {
            return 0;
        }
    }

    public NSString capitalizeStringDynamic() {
        StringBuilder res = new StringBuilder();
        boolean capitalizeNext = true;
        for (char c : string.toCharArray()) {
            if (Character.isWhitespace(c)) {
                capitalizeNext = true;
                res.append(c);
            } else if (capitalizeNext) {
                res.append(Character.toUpperCase(c));
                capitalizeNext = false;
            } else {
                res.append(c);
            }
        }
        return new NSString(res.toString());
    }

    public NSString lowercaseStringDynamic() {
        return new NSString(string.toLowerCase(Locale.ROOT));
    }

    public NSString uppercaseStringDynamic() {
        return new NSString(string.toUpperCase(Locale.ROOT));
    }

    public NSString stringByAppendingStringDynamic(NSString other) {
        return new NSString(string + other.string);
    }

    public boolean isEqualToStringDynamic(NSString other) {
        return this.string.equals(other.string);
    }

    @Override
    public boolean isEqualDynamic(NSObject other) {
        if (!(other instanceof NSString)) return false;
        NSString strObj = (NSString) other;
        return this.string.equals(strObj.string);
    }

    @Override
    public NSString initDynamic() {
        super.initDynamic();
        return this;
    }

    @Override
    public String descriptionDynamic() {
        return string;
    }

    // ------------------- Инициализация ClassInfo -------------------
    public static void initClassInfo(ClassInfo info) {
        s_classInfo = info;
    }

    // ------------------- Getter для строки -------------------
    public String getString() {
        return string;
    }
}

File: rtl/NSString.cpp
#include "NSString.h"
#include <algorithm>
#include <cctype>
#include <cstring>
#include <iostream>

// Статический ClassInfo
ClassInfo* NSString::s_classInfo = nullptr;

// ------------------- Конструкторы -------------------
NSString::NSString() : string("") {}
NSString::NSString(const std::string& str) : string(str) {}
NSString::NSString(const NSString& other) : string(other.string) {}

// ------------------- Статические методы -------------------
std::unique_ptr<NSString> NSString::stringStatic() {
    return std::make_unique<NSString>("");
}

std::unique_ptr<NSString> NSString::stringWithCStringStatic(const char* cstr) {
    return std::make_unique<NSString>(std::string(cstr));
}

std::unique_ptr<NSString> NSString::stringWithStringStatic(const NSString& str) {
    return std::make_unique<NSString>(str);
}

// ------------------- Динамические методы -------------------
std::unique_ptr<char[]> NSString::cStringDynamic() const {
    auto res = std::make_unique<char[]>(string.size() + 1);
    for (size_t i = 0; i < string.size(); ++i) {
        res[i] = string[i];
    }
    res[string.size()] = '\0';
    return res;
}


std::unique_ptr<NSString> NSString::capitalizeStringDynamic() {
    std::string res = string;
    bool capitalizeNext = true;
    for (char& c : res) {
        if (isspace(static_cast<unsigned char>(c))) {
            capitalizeNext = true;
        } else if (capitalizeNext) {
            c = static_cast<char>(std::toupper(c));
            capitalizeNext = false;
        }
    }
    return std::make_unique<NSString>(res);
}

char NSString::characterAtIndexDynamic(int index) {
    return (index >= 0 && index < static_cast<int>(string.size())) ? string[index] : '\0';
}

int NSString::hasPrefixDynamic(const NSString& prefix) {
    return string.rfind(prefix.string, 0) == 0 ? 1 : 0;
}

int NSString::hasSuffixDynamic(const NSString& suffix) {
    if (suffix.string.size() > string.size()) return 0;
    return string.compare(string.size() - suffix.string.size(), suffix.string.size(), suffix.string) == 0 ? 1 : 0;
}

std::unique_ptr<NSString> NSString::init() {
    NSObject::initDynamic();
    return std::make_unique<NSString>(*this);
}

int NSString::intValueDynamic() {
    return std::stoi(string);
}

int NSString::isEqualDynamic(NSObject* other) {
    auto strObj = dynamic_cast<NSString*>(other);
    if (!strObj) return 0;
    return string == strObj->string ? 1 : 0;
}

int NSString::isEqualToStringDynamic(const NSString& other) {
    return string == other.string ? 1 : 0;
}

int NSString::lengthDynamic() {
    return static_cast<int>(string.size());
}

std::unique_ptr<NSString> NSString::lowercaseStringDynamic() {
    std::string res = string;
    std::transform(res.begin(), res.end(), res.begin(), ::tolower);
    return std::make_unique<NSString>(res);
}

std::unique_ptr<NSString> NSString::uppercaseStringDynamic() {
    std::string res = string;
    std::transform(res.begin(), res.end(), res.begin(), ::toupper);
    return std::make_unique<NSString>(res);
}

std::unique_ptr<NSString> NSString::stringByAppendingStringDynamic(const NSString& other) {
    return std::make_unique<NSString>(string + other.string);
}

// ⚡ Виртуальный метод базового класса NSObject
std::string NSString::descriptionDynamic() {
    return string;
}

// ------------------- Инициализация ClassInfo -------------------
void NSString::initClassInfo(ClassInfo* info) {
    s_classInfo = info;
}
